### PR TITLE
Add doxygen documentation

### PR DESCRIPTION
**Pull request by justwagle**
Adding basic infrastructure to generate doxygen documentation.

### CODE REVIEW

File: cmake/Phylanx_Documentation.cmake [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# Copyright (c) 2018 Bibek Wagle
+#
+# Distributed under the Boost Software License, Version 1.0. (See accompanying
+# file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+option(PHYLANX_BUILD_DOCS "Build Phylanx Documentation" OFF)
```
Comment (hkaiser) [inline cmake/Phylanx_Documentation.cmake]: You might want to consider using `phylanx_option()` instead (see here for some examples: https://github.com/STEllAR-GROUP/phylanx/blob/master/CMakeLists.txt#L86-L104).

File: docs/Doxyfile [hunk 2]
```diff
@@ -0,0 +1,2494 @@
+# Doxyfile 1.8.13
```
Comment (hkaiser) [inline docs/Doxyfile]: I'd suggest not to blindly generate the docs from all files in the project but rather add the files that actually contain doxygen comments manually one by one to the doxygen invocation. Having blindly generated docs is unneededly verbose and essentially useless. See the HPX documentation toolchain for examples (https://github.com/STEllAR-GROUP/hpx/blob/master/docs/CMakeLists.txt)

File: docs/Doxy.in [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+EXTRACT_ALL = YES
+GENERATE_HTML = YES
+MACRO_EXPANSION = YES
+PREDEFINED = @DOCYGEN_DEFINATIONS@
+OUTPUT_DIRECTORY = @CMAKE_CURRENT_BINARY_DIR@/doxygen_API_docs/
+GENERATE_LATEX = YES
+HAVE_DOT = NO
+INPUT = @DOXYGEN_INPUTS@
+ENABLE_PREPROCESSING = NO
+SKIP_FUNCTION_MACROS = YES
```
Comment (hkaiser) [inline docs/Doxy.in]: This file does not have a end of line character at the end of the last line.

Comment (hkaiser) [inline docs/Doxy.in]: Could you put this template file into `$PROJECT_ROOT/cmake/templates/`, please?

Comment (hkaiser) [inline docs/Doxy.in]: `EXCLUDE_SYMBOLS = detail` might be a good choice to add as it automatically skips everything in `namespace detail`.

**Comment by hkaiser** [comment]
Could you please merge all the commits into one? That would avoid for the large file to end up in the repository history for no good reason.

**Comment by justwagle** [comment]
@hkaiser I consolidated the commits and forced pushed a clean history and also addressed the review comments.

**Pull request merged**